### PR TITLE
Scroll

### PR DESCRIPTION
**Pull request by nitishgoyal13**

### CODE REVIEW

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java [hunk 1]
```diff
@@ -597,6 +598,7 @@ public void testQueryMultipleFiltersOrCombiner() throws FoxtrotException, JsonPr
     }
 
     @Test
```
Comment (nitishgoyal13) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java:600]: Pagination is not supported anymore
https://www.elastic.co/guide/en/elasticsearch/reference/6.8/breaking-changes-6.0.html#_scroll

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java [hunk 2]
```diff
@@ -749,11 +751,10 @@ public void testQueryCaching() throws FoxtrotException, JsonProcessingException
         equalsFilter.setValue("ios");
         query.setFilters(Lists.<Filter>newArrayList(equalsFilter));
 
-        query.setFrom(1);
         query.setLimit(1);
 
         List<Document> documents = new ArrayList<Document>();
```
Comment (nitishgoyal13) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java:756]: https://www.elastic.co/guide/en/elasticsearch/reference/6.8/breaking-changes-6.0.html#_scroll

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java [hunk 3]
```diff
@@ -107,22 +112,23 @@ public void validateImpl(Query parameter) {
             validationErrors.add("limit must be positive integer");
         }
 
+        if (parameter.getLimit() >= elasticsearchTuningConfig.getDocumentsLimitAllowed()){
```
Comment (jitendradhawan) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java:115]: greater than only

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/Query.java [hunk 4]
```diff
@@ -37,6 +39,9 @@
 
     private int limit = 10;
 
+    private boolean scrollRequest = false;
```
Comment (jitendradhawan) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/Query.java:42]: we can separate out scroll query request in a new concrete action request class

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java [hunk 5]
```diff
@@ -145,13 +151,108 @@ public SearchRequest getRequestBuilder(Query parameter) {
...
+                .indicesOptions(Utils.indicesOptions())
+                .types(ElasticsearchUtils.DOCUMENT_TYPE_NAME)
+                .searchType(SearchType.QUERY_THEN_FETCH)
+                .source(new SearchSourceBuilder()
+                                .timeout(new TimeValue(getGetQueryTimeout(), TimeUnit.MILLISECONDS))
+                                .size(parameter.getLimit())
+                                .query(new ElasticSearchQueryGenerator().genFilter(parameter.getFilters()))
+                                .sort(Utils.storedFieldName(parameter.getSort().getField()),
+                                      ResultSort.Order.desc == parameter.getSort().getOrder()
+                                      ? SortOrder.DESC : SortOrder.ASC));
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java:185]: Refactor to generate `SearchSourceBuilder` only in one place

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java [hunk 6]
```diff
@@ -107,22 +112,23 @@ public void validateImpl(Query parameter) {
...
-            SearchResponse response = getConnection()
-                .getClient()
-                .search(search);
-            return getResponse(response, parameter);
-        } catch (IOException e) {
-            throw FoxtrotExceptions.createQueryExecutionException(parameter, e);
+        if (parameter.isScrollRequest()) {
+            return executeScrollRequest(parameter);
+        }
+        else {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java:130]: Remove else and let this be the default path. Looks cleaner. Compiler will add the else for you.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java [hunk 7]
```diff
@@ -145,13 +151,108 @@ public SearchRequest getRequestBuilder(Query parameter) {
...
+        }
+    }
+
+    private QueryResponse getResponse(Query parameter, long totalHits, List<String> ids, String scrollId) {
+        if (ids.isEmpty()) {
+            return QueryResponse
+                    .builder()
+                    .documents(Collections.emptyList())
+                    .totalHits(0)
+                    .scrollResponse(true)
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java:232]: what makes sense is to remove this variable and create one called moreDataAvailable. Set it to false in this case, true in the other.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java [hunk 8]
```diff
@@ -145,13 +151,108 @@ public SearchRequest getRequestBuilder(Query parameter) {
...
+                    .build();
+        }
+        return QueryResponse
+                .builder()
+                .documents(getQueryStore()
+                                   .getAll(parameter.getTable(), ids, true))
+                .totalHits(totalHits)
+                .scrollId(scrollId)
+                .scrollResponse(true)
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/FilterAction.java:241]: Set moreDataAvailable true in this case

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/MetricRecorder.java [hunk 9]
```diff
@@ -11,7 +11,7 @@
 public class MetricRecorder implements ActionExecutionObserver {
     @Override
     public void postExecution(ActionEvaluationResponse response) {
-        if(null == response.getExecutedAction()) {
+        if(response == null  || null == response.getExecutedAction()) {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/MetricRecorder.java:14]: follow same pattern for null check

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java [hunk 10]
```diff
@@ -33,6 +33,7 @@
 import com.flipkart.foxtrot.core.exception.FoxtrotException;
 import com.flipkart.foxtrot.core.querystore.impl.ElasticsearchUtils;
 import com.google.common.collect.Lists;
+import com.sun.tools.javac.util.Assert;
```
Comment (santanusinha) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java:36]: This is wrong import. Use junit Assert.

**Pull request merged**